crates/precompile/src/bn128/arkworks.rs

```diff
@@ -0,0 +1,214 @@
+use super::{FQ2_LEN, FQ_LEN, G1_LEN, SCALAR_LEN};
+use crate::PrecompileError;
+use std::vec::Vec;
+
+use ark_bn254::{Bn254, Fq, Fq2, Fr, G1Affine, G1Projective, G2Affine};
+use ark_ec::{pairing::Pairing, AffineRepr, CurveGroup};
+use ark_ff::{One, PrimeField, Zero};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+
+/// Reads a single `Fq` field element from the input slice.
+///
+/// Takes a byte slice and attempts to interpret the first 32 bytes as an
+/// elliptic curve field element. Returns an error if the bytes do not form
+/// a valid field element.
+///
+/// # Panics
+///
+/// Panics if the input is not at least 32 bytes long.
+#[inline]
+fn read_fq(input_be: &[u8]) -> Result<Fq, PrecompileError> {
+    assert_eq!(input_be.len(), FQ_LEN, "input must be {FQ_LEN} bytes");
+
+    let mut input_le = [0u8; FQ_LEN];
+    input_le.copy_from_slice(input_be);
+
+    // Reverse in-place to convert from big-endian to little-endian.
+    input_le.reverse();
+
+    Fq::deserialize_uncompressed(&input_le[..])
+        .map_err(|_| PrecompileError::Bn128FieldPointNotAMember)
+}
+/// Reads a Fq2 (quadratic extension field element) from the input slice.
+///
+/// Parses two consecutive Fq field elements as the real and imaginary parts
+/// of an Fq2 element.
+/// The second component is parsed before the first, ie if a we represent an
+/// element in Fq2 as (x,y) -- `y` is parsed before `x`
+///
+/// # Panics
+///
+/// Panics if the input is not at least 64 bytes long.
+#[inline]
+fn read_fq2(input: &[u8]) -> Result<Fq2, PrecompileError> {
+    let y = read_fq(&input[..FQ_LEN])?;
+    let x = read_fq(&input[FQ_LEN..2 * FQ_LEN])?;
+
+    Ok(Fq2::new(x, y))
+}
+
+/// Creates a new `G1` point from the given `x` and `y` coordinates.
+///
+/// Constructs a point on the G1 curve from its affine coordinates.
+///
+/// Note: The point at infinity which is represented as (0,0) is
+/// handled specifically because `AffineG1` is not capable of
+/// representing such a point.
+/// In particular, when we convert from `AffineG1` to `G1`, the point
+/// will be (0,0,1) instead of (0,1,0)
+#[inline]
+fn new_g1_point(px: Fq, py: Fq) -> Result<G1Affine, PrecompileError> {
+    if px.is_zero() && py.is_zero() {
+        Ok(G1Affine::zero())
+    } else {
+        let point = G1Affine::new_unchecked(px, py);
```

Comment:
```suggestion
        // We cannot use `G1Affine::new` because that triggers an assert if the point is not on the curve.
        let point = G1Affine::new_unchecked(px, py);
```